### ISSUE TITLE
Rule enhancement: More granular control in space-unary-ops

### ISSUE THREAD

**Issue by Alaneor** [eslint/eslint#5060]
### Current state

This rule allows us to configure whether we expect spaces between unary operators for either word operators (`delete`, `void`, `typeof` etc.) and non-word (`+`, `-`, `++`, `!`, etc.).

### Problem

Some might prefer different spacing rules for individual non-word operators. For me personally, I prefer to have the negation operator `!` always separated from its operand, but I never separate `++` or `--` from its operand.

### Proposal

Extend the rule's configuration so that we can configure the spacing rules for individual operators. One example might be to add an object as third argument where the object's key would be the operator and the key's value would be the rule to apply:

```js
{
  'space-unary-ops': [1, {
    words: false,
    nonwords: false,
    '!': true  // Overrides spacing rule for ! to always have space
  }]
}
```

> The configuration scheme above is just an idea. The goal is to allow more fine-grained control over where the spaces are expected and where they are not.

Thanks!

**Comment by eslintbot** [comment]
@Alaneor Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by ilyavolodin** [comment]
Sorry, it looks like this issue slipped through the cracks. I don't think I like proposed schema, since it might be very confusing, but maybe creating 3-rd parameter `overrides` which could accept an object might be fine. @eslint/eslint-team any problems with marking this as accepted?

**Comment by nzakas** [comment]
`overrides` seems like a good way to solve it.

**Comment by ilyavolodin** [comment]
So proposed schema will look like this:
```js
{
  'space-unary-ops': [1,
    words: false,
    nonwords: false,
    { overrides: {
      '!': true  // Overrides spacing rule for ! to always have space
    }
  }
  }]
}
```
Slightly more verbose then I'd like, but more descriptive, and less confusing, I think.

**Comment by Alaneor** [comment]
That looks good to me. As long as I can override individual operands. :)

The only argument against a third `overrides` property that comes to mind is that I don't think I have seen such option on any other rule. That was the reason I proposed the configuration to be a single object.

**Comment by ilyavolodin** [comment]
We have a few rules that do that:
http://eslint.org/docs/rules/keyword-spacing.html#options
http://eslint.org/docs/rules/operator-linebreak.html#ignore-override

**Comment by Alaneor** [comment]
Oh, I see! I totally missed that.

**Comment by nzakas** [comment]
@Alaneor do you want to submit a pull request?

**Comment by afahim** [comment]
@Alaneor (or perhaps @ilyavolodin): can you provide an example where a space would not be expected after a word operator?

**Comment by ilyavolodin** [comment]
@afahim The only one that comes to my mind is:
```js
typeof(foo) === 'blah';
```

**Comment by platinumazure** [comment]
@afahim `typeof` and `delete` were the only word unary operators I could come up with. They both can be followed by parentheses to avoid spacing (unless the rule is turned on to enforce spacing).

Additionally, it might be worth seeing if `no-extra-parens` would warn in those cases-- maybe in the case where the operand is in parentheses but there is no space, `no-extra-parens` should not warn.

**Comment by afahim** [comment]
Cool, thanks guys. I was wondering if I were missing any cases besides the parenthesis with expressions.

I'll take a look at this during the weekend.

**Comment by Alaneor** [comment]
Thanks to everyone who worked on this! Great work!

**Comment by tomas-sentkeresty** [comment]
Hello, what if, we want to enforce usage:
```js
typeof(foo) === 'blah';
```
instead of
```js
typeof (foo) === 'blah';
```
Second case should end up with error.
Thank you.
